virttest/qemu_vm.py

```diff
@@ -1484,6 +1484,26 @@ def sort_key(dev):
         if params.get("defconfig", "yes") == "no":
             devices.insert(StrDev('nodefconfig', cmdline=" -nodefconfig"))
 
+        # Add device intel-iommu, it must be added before any virtio devices
+        if (params.get('intel_iommu') == 'yes' and
+                devices.has_device('intel-iommu')):
+            iommu_params = {
+                'intremap': params.get('iommu_intremap', 'on'),
+                'device-iotlb': params.get('iommu_device-iotlb', 'on')}
+            iommu_dev = QDevice('intel-iommu', iommu_params)
+            iommu_dev.set_param(
+                'caching-mode', params.get('iommu_caching_mode'))
```

Comment:
@vivianQizhu  Both can work, but could you explain why don't set all the params first then create QDevice ?

Reply:
@yanan-fu Because `set_param` can handle it when certain parameter is `None`, to define them all before create QDevice might hit error if the user does not specify all of them.

Reply:
Qdevice use `set_param` too ?
https://github.com/avocado-framework/avocado-vt/blob/2cc74315c7eb4f05d052428e79a4f4739754a601/virttest/qemu_devices/qdevices.py#L89

I tested with set all params before create QDevice, it can pass too.

Reply:
If it is not must, i prefer use same method for all the parameter a bit more ~